### PR TITLE
add facet support

### PR DESCRIPTION
**Pull request by tonysun83**
We add support for counts, ranges, and drilldown.

### CODE REVIEW

File: src/search3_rpc.erl [hunk 1]
```diff
@@ -242,6 +248,34 @@ construct_sort_msg(SortArg) when is_list(SortArg) ->
...
+    maps:from_list(RangesList);
+construct_ranges_msg(_) ->
+    #{}.
+
+construct_drilldown_msg([]) ->
+    [];
+construct_drilldown_msg(Drilldown) when is_list(Drilldown) ->
+    [#{parts => Part} || Part <- Drilldown];
+construct_drilldown_msg(_) ->
+    [].
```
Comment (rnewson) [inline src/search3_rpc.erl]: is this a defensive programming clause? If so, best remove. the crash is better.

File: src/search3_rpc.erl [hunk 2]
```diff
@@ -242,6 +248,34 @@ construct_sort_msg(SortArg) when is_list(SortArg) ->
...
+    MakeRangeFun = fun
+        ({K, {V}}) ->
+            SubList = [{?b2l(F), ?b2l(R)} || {F, R} <- V],
+            SubMap = maps:from_list(SubList),
+            {?b2l(K), #{ranges => SubMap}}
+    end,
+    RangesList = lists:map(MakeRangeFun, Ranges),
+    maps:from_list(RangesList);
+construct_ranges_msg(_) ->
+    #{}.
```
Comment (rnewson) [inline src/search3_rpc.erl]: is this a defensive programming clause? If so, best remove. the crash is better.

File: src/search3_rpc.erl [hunk 3]
```diff
@@ -242,6 +248,34 @@ construct_sort_msg(SortArg) when is_list(SortArg) ->
 construct_sort_msg(_) ->
     #{}.
 
+construct_counts_msg(Counts) when is_binary(Counts) ->
+    [Counts];
+construct_counts_msg(Counts) when is_list(Counts) ->
+    Counts;
+construct_counts_msg(_) ->
+    [].
```
Comment (rnewson) [inline src/search3_rpc.erl]: is this a defensive programming clause? If so, best remove. the crash is better.

File: src/search3_rpc.erl [hunk 4]
```diff
@@ -242,6 +248,34 @@ construct_sort_msg(SortArg) when is_list(SortArg) ->
 construct_sort_msg(_) ->
     #{}.
```
Comment (rnewson) [inline src/search3_rpc.erl]: is this a defensive programming clause? If so, best remove. the crash is better.

File: src/search3_response.erl [hunk 5]
```diff
@@ -65,6 +66,15 @@ order_to_json(Order) ->
...
 
+facets_to_json(Counts) ->
+    CountsList = maps:to_list(Counts),
+    CountsFun = fun
+        ({K, V}) when is_binary(K), is_map(V) ->
+            #{counts := SubCounts} = V,
+            SubCountsList = maps:to_list(SubCounts),
+            {K, {SubCountsList}}
+    end,
+    {lists:map(CountsFun, CountsList)}.
```
Comment (rnewson) [inline src/search3_response.erl:77]: not sure why we turn the map into a list when jiffy can handle maps?

```
jiffy:encode(#{<<"foo">> => <<"bar">>}).
<<"{\"foo\":\"bar\"}">>
```

Comment (tonysun83) [reply src/search3_response.erl:77]: @rnewson: If we did a jiffy:encode, the binary format
`<<"{\"class\":{\"counts\":{\"mammal\":8,\"bird\":2}}}">>` is returned to the user in readable format but it breaks existing testy tests that expects a nested object. 

The list conversion creates the following nested object:
```
    "counts": {
        "class": {
            "bird": 2,
            "mammal": 8
        }
    }
```

Comment (rnewson) [reply src/search3_response.erl:77]: fair enough.

**Review by rnewson — APPROVED** [review]
+1

**Pull request merged**